tests/devices/test_caching.py

```diff
@@ -0,0 +1,277 @@
+            qml.QuantumFunctionError, match="Device caching is incompatible with the backprop"
+        ):
+            QNode(qfunc, dev, diff_method="backprop")
+
+    def test_gradient_autograd(self, mocker):
+        """Test that caching works when calculating the gradient method using the autograd
+        interface"""
+        dev = qml.device("default.qubit", wires=2, caching=10)
+        qn = QNode(qfunc, dev, interface="autograd")
+        d_qnode = qml.grad(qn)
+        args = [0.1, 0.2]
+
+        d_qnode(*args)
+        spy = mocker.spy(DefaultQubit, "apply")
+        d_qnode(*args)
+        spy.assert_not_called()
+
+    @pytest.mark.usefixtures("skip_if_no_tf_support")
+    def test_gradient_tf(self, mocker):
+        """Test that caching works when calculating the gradient method using the TF interface"""
+        import tensorflow as tf
+
+        dev = qml.device("default.qubit", wires=2, caching=10)
+        qn = QNode(qfunc, dev, interface="tf")
+        args0 = tf.Variable(0.1)
+        args1 = tf.Variable(0.2)
+
+        with tf.GradientTape() as tape:
+            res = qn(args0, args1)
+
+        grad = tape.gradient(res, args0)
+        assert grad is not None
+
+        spy = mocker.spy(DefaultQubit, "apply")
+        with tf.GradientTape() as tape:
+            res = qn(args0, args1)
+
+        tape.gradient(res, args0)
+        spy.assert_not_called()
+
+    @pytest.mark.usefixtures("skip_if_no_torch_support")
+    def test_gradient_torch(self, mocker):
+        """Test that caching works when calculating the gradient method using the Torch interface"""
+        import torch
+
+        dev = qml.device("default.qubit", wires=2, caching=10)
+        qn = QNode(qfunc, dev, interface="torch")
+        args0 = torch.tensor(0.1, requires_grad=True)
+        args1 = torch.tensor(0.2)
+
+        res = qn(args0, args1)
+        res.backward()
+        assert args0.grad is not None
+
+        spy = mocker.spy(DefaultQubit, "apply")
+        res = qn(args0, args1)
+        res.backward()
+        spy.assert_not_called()
+
+    def test_mutable_circuit(self, mocker):
+        """Test that caching is compatible with circuit mutability. Caching should take place if
+        the circuit and parameters are the same, and should not take place if the parameters are
+        the same but the circuit different."""
+        dev = qml.device("default.qubit", wires=3, caching=10)
+
+        @qnode(dev)
+        def qfunc(x, y, flag=1):
+            if flag == 1:
+                qml.RX(x, wires=0)
+                qml.RX(y, wires=1)
+            else:
+                qml.RX(x, wires=1)
+                qml.RX(y, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return expval(qml.PauliZ(wires=1))
+
+        spy = mocker.spy(DefaultQubit, "apply")
+        qfunc(0.1, 0.2)
+        qfunc(0.1, 0.2)
+        assert len(spy.call_args_list) == 1
+
+        qfunc(0.1, 0.2, flag=0)
+        assert len(spy.call_args_list) == 2
```

Comment:
I guess we really have no way of testing that caching is not used when there are classical processing steps inside the qfunc which are not deterministic given the inputs?

Reply:
That could be tested as well, no?

Reply:
I wasn't thinking of "tests", I was more thinking that this is a case we can't support properly in the software itself, since there's no way to distinguish such functions from inspecting their signatures and ops?

Reply:
Oh. Yeah, you're right. A QNode that contains non-deterministic classical processing will simply result in a new tape, and thus new hash, every evaluation. So even if caching is turned on, it will probably never encounter two tapes with the same hash.